src/auth/webhook-secret.guard.ts

```diff
@@ -0,0 +1,32 @@
+import {
+  CanActivate,
+  ExecutionContext,
+  Injectable,
+  UnauthorizedException,
+} from '@nestjs/common';
+import * as crypto from 'crypto';
+import { Request } from 'express';
+import { Env } from 'src/env';
+
+@Injectable()
+export class WebhookSecretGuard implements CanActivate {
+  canActivate(context: ExecutionContext): boolean {
+    const request: Request = context.switchToHttp().getRequest();
+    const signature = request.headers['x-hub-signature-256'] as string;
+    const secret = Env.SECRET.WEBHOOK;
+    const payload = JSON.stringify(request.body);
+
+    if (!signature) {
+      throw new UnauthorizedException('Missing signature');
+    }
+
+    const hmac = crypto.createHmac('sha256', secret);
+    const digest = `sha256=${hmac.update(payload).digest('hex')}`;
+
+    if (!crypto.timingSafeEqual(Buffer.from(digest), Buffer.from(signature))) {
+      throw new UnauthorizedException('Invalid signature');
+    }
+
+    return true;
+  }
```

Comment:
PullProofAI

The `WebhookSecretGuard` class looks well-implemented for verifying webhook signatures. However, there are a few improvements that can be suggested:

1. **Error Handling**: It's good to throw exceptions for missing or invalid signatures, but consider providing more specific error messages for better debugging and security.
2. **Code Readability**: Break down the logic into smaller functions for better readability and maintainability.
3. **Type Safety**: Ensure proper type annotations for variables and function return types.

Consider the following suggestions for improvement:

```diff
  canActivate(context: ExecutionContext): boolean {
    const request: Request = context.switchToHttp().getRequest();
    const signature = request.headers['x-hub-signature-256'] as string;
    const secret = Env.SECRET.WEBHOOK;
    const payload = JSON.stringify(request.body);

    if (!signature) {
      throw new UnauthorizedException('Missing signature in the request headers');
    }

    const isValidSignature = this.validateSignature(payload, signature, secret);

    if (!isValidSignature) {
      throw new UnauthorizedException('Invalid signature in the request headers');
    }

    return true;
  }
+
+   private validateSignature(payload: string, signature: string, secret: string): boolean {
+     const hmac = crypto.createHmac('sha256', secret);
+     const digest = `sha256=${hmac.update(payload).digest('hex')}`;
+ 
+     return crypto.timingSafeEqual(Buffer.from(digest), Buffer.from(signature));
+   }
```

These changes enhance error handling, readability, and maintainability of the `WebhookSecretGuard` class.


<!-- This is an auto-generated comment by PullProofAI -->